Add utilize eye in hand calibration sample

chrisasc opened the pull request:
Add new sample in C# similar to the one we have in C++ (https://github.com/zivid/zivid-cpp-samples/blob/master/source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.cpp)

Im not sure if I added the dependencies correctly. Now they are inside the folder "Packages" in the sample folder...

SatjaSivcev commented on README.md:
```diff
@@ -30,7 +30,12 @@ There are two main categories of samples: **Camera** and **Applications**. The s
       - [**ZDF2PLY**][ZDF2PLY-url] - Convert point cloud from ZDF a file to a PLY file.
   - **Advanced**
-    - [**HandEyeCalibration**][HandEyeCalibration-url] - Perform Hand-Eye calibration
+  - [**HandEyeCalibration**][HandEyeCalibration-url]
+      - [**HandEyeCalibration**][HandEyeCalibrationSample-url] - Perform Hand-Eye calibration.
+      - [**UtilizeEyeInHandCalibration**][UtilizeEyeInHandCalibration-url] - Transform single data point or entire point cloud from camera frame to robot base frame using Eye-in-Hand calibration matrix.
+	  - **Dependencies:**
+		  - [Math.NET Numerics](https://numerics.mathdotnet.com/#Math-NET-Numerics) version 4.12.0 or newer
+          - [Yaml.NET](https://github.com/aaubry/YamlDotNet/wiki) version 8.1.2 or newer
```
HandEyeCalibration and all below needs to be indented for one step

![image](https://user-images.githubusercontent.com/46567886/91553362-8f961800-e92d-11ea-9224-cfa271044236.png)

chrisasc replied:
done

SatjaSivcev commented on README.md:
```diff
@@ -30,7 +30,12 @@ There are two main categories of samples: **Camera** and **Applications**. The s
+      - [**HandEyeCalibration**][HandEyeCalibrationSample-url] - Perform Hand-Eye calibration.
+      - [**UtilizeEyeInHandCalibration**][UtilizeEyeInHandCalibration-url] - Transform single data point or entire point cloud from camera frame to robot base frame using Eye-in-Hand calibration matrix.
+	  - **Dependencies:**
+		  - [Math.NET Numerics](https://numerics.mathdotnet.com/#Math-NET-Numerics) version 4.12.0 or newer
+          - [Yaml.NET](https://github.com/aaubry/YamlDotNet/wiki) version 8.1.2 or newer
```
There are no instructions on how to install/download or whatever needs to be done using these dependencies.

I suppose we should include instructions on: https://zivid.atlassian.net/wiki/spaces/ZividKB/pages/534708242/Build+C+Samples+using+Visual+Studio

In addition, the readme file should have a link to the above article. Same as we have for C++ see:
![image](https://user-images.githubusercontent.com/46567886/91553916-7c377c80-e92e-11ea-95df-7687135c6900.png)

chrisasc replied:
Updated readme to be more similar like the C++ readme

SatjaSivcev replied:
Should we add this link to our instructions? https://docs.microsoft.com/en-us/nuget/consume-packages/install-use-packages-powershell
I didin't know what package manager console is.

chrisasc replied:
It is something that is inside Visual Studio, but I can explain how to open it and if they dont have it, install it using this link. (Better to be safe the sorry)

knatten replied:
I don't think we need any instructions for this, since it happens automatically during build. What do you think?

SatjaSivcev commented on README.md:
```diff
@@ -66,5 +71,7 @@ Zivid Samples are distributed under the [BSD license](LICENSE).
 [ZDF2PLY-url]: source/Applications/Basic/FileFormats/ZDF2PLY/ZDF2PLY.cs
-[HandEyeCalibration-url]: source/Applications/Advanced/HandEyeCalibration/HandEyeCalibration.cs
+[HandEyeCalibration-url]: source/Applications/Advanced/HandEyeCalibration/README.md
+[HandEyeCalibrationSample-url]: source/Applications/Advanced/HandEyeCalibration/HandEyeCalibration/HandEyeCalibration.cs
```
Call this one HandEyeCalibration-url and the main one HandEyeCalibrationReadme-url?

SatjaSivcev replied:
If you change this, make a small correction PR for C++

chrisasc replied:
done

SatjaSivcev commented on source/Applications/Advanced/HandEyeCalibration/README.md:
```diff
@@ -0,0 +1,34 @@
+
+[ZividHandEyeCalibration][ZividHandEyeCalibration-url]
+
+* CLI application which takes a collection of pose pairs (e.g. output of steps 1.-3. in [HandEyeCalibration][HandEyeCalibration-url]) and returns a **Transformation Matrix**
```
steps are here arabic (1. - 3.) but up there they are roman (i. - iii.)

![image](https://user-images.githubusercontent.com/46567886/91555948-354b8600-e932-11ea-88ed-b534da93afc5.png)

SatjaSivcev replied:
If you change this, make a small correction PR for C++

chrisasc replied:
I cannot change this because they way of numbering is set in Github's CSS.  In Visual Studio Code the markdown text is displayed in Arabic. So we must just let it be... 

Found the answer from this link:
https://github.community/t/github-ordered-list-is-displayed-in-roman-numbers/119391

SatjaSivcev replied:
You can change the arabic to be roman (the hard-coded part)?

chrisasc replied:
yes, updated it now. Done

SatjaSivcev commented on source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.csproj:
```diff
@@ -0,0 +1,91 @@
+      <HintPath>$(ZIVID_INSTALL_FOLDER)\bin_debug\ZividCoreNET.dll</HintPath>
+    </Reference>
+    <Reference Include="ZividVisualizationNET">
+      <HintPath>..\..\..\..\..\..\..\..\Program Files\Zivid\bin\ZividVisualizationNET.dll</HintPath>
```
Use variable for path. See https://github.com/zivid/zivid-csharp-samples/blob/add-utilize-eye-in-hand-calibration-sample/source/Applications/Basic/Visualization/CaptureVis3D/CaptureVis3D.csproj

chrisasc replied:
done

SatjaSivcev commented on source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/packages/YamlDotNet.8.1.2/LICENSE.txt:
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2008, 2009, 2010, 2011, 2012, 2013, 2014 Antoine Aubry and contributors
```
I am not sure whether we want to host the packages in the repo. SE needs to be involved. Even if we do, this is probably not the right directory.

chrisasc replied:
Removing the packages! Instead, Im explaining in the readme how to install them

SatjaSivcev commented on source/ZividNETSamples.sln:
```diff
@@ -1,7 +1,7 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.28307.902
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.30320.27
```
What version of Visual Studio do you use? 2019?

chrisasc replied:
Yes, Im using 2019 now

chrisasc replied:
No, it is probably better to keep it to a constant version. I have changed it back to the original one (VS 2015)

SatjaSivcev commented on source/ZividNETSamples.sln:
```diff
@@ -31,80 +31,122 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CaptureFromFileVis3D", "App
+EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|Any CPU = Debug|Any CPU
```
Why is this added?

chrisasc replied:
I dont know

SatjaSivcev replied:
Maybe it depends on how your Visual Studio is configured. Do you know @knatten

knatten replied:
> I dont know, but is is also included in the current one. So I'm assuming that is how it is suppose to be

What do you mean? I don't see "Any CPU" configurations anywhere on master?

chrisasc replied:
I was misunderstanding Satjas question. They are added because of my configurations in Visual Studio, updated it now to exclude the "Any CPU" stuff

SatjaSivcev commented on source/ZividNETSamples.sln:
```diff
@@ -31,80 +31,122 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CaptureFromFileVis3D", "App
 		Release|x64 = Release|x64
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
+		{DD533F88-E34D-41A8-B734-1B94A5A4A31F}.Debug|Any CPU.ActiveCfg = Debug|x64
```
Why are all these added?

chrisasc replied:
no idea. the snl files is something that is automatically created

SatjaSivcev replied:
Maybe it depends on how your Visual Studio is configured. Do you know @knatten

chrisasc replied:
It happens because I have not configured Visual Studio correctly. The default for Visual Studio is "Any CPU", so I changed this and now I dont generate .snl files with all these "Any CPU" stuff

SatjaSivcev commented on source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.csproj:
```diff
@@ -0,0 +1,91 @@
+    <Reference Include="System.Xml" />
+    <Reference Include="WindowsBase" />
+    <Reference Include="YamlDotNet, Version=8.0.0.0, Culture=neutral, PublicKeyToken=ec19458f3c15af5e, processorArchitecture=MSIL">
+      <HintPath>packages\YamlDotNet.8.1.2\lib\net45\YamlDotNet.dll</HintPath>
```
This has to be removed in case we decide not to have packages in the repo.

chrisasc replied:
removed

SatjaSivcev commented on source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.csproj:
```diff
@@ -0,0 +1,91 @@
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="MathNet.Numerics, Version=4.12.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>packages\MathNet.Numerics.4.12.0\lib\net461\MathNet.Numerics.dll</HintPath>
```
This has to be removed if we decide not to have packages in the repo.

chrisasc replied:
removed

SatjaSivcev commented on source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.cs:
```diff
@@ -0,0 +1,157 @@
+
+                case CommandType.TransformPointCloud:
+
+                    Console.WriteLine("Transforming point cloud");
+                    var transformMatrix = mathDotNetToZivid(transformBaseToCamera);
+                    pointCloud.Transform(transformMatrix);
+
+                    var saveFile = "ZividGemTransformed.zdf";
+                    Console.WriteLine("Saving frame to file: " + saveFile);
+                    frame.Save(saveFile);
```
Should we add a comment that calling transform() on pointCloud actually changes the frame?

knatten replied:
I don't have a strong opinion on that. In my opinion we do not to explain all the api calls in samples, we should only do it if it's of some particular interest in a particular sample. Whether that applies here, I'm not too sure about, I think that's more up to what you want to communicate with this sample.

SatjaSivcev replied:
I would not expect that calling transform would modify the frame. However, the sample does show exactly that and the file name indicates it so maybe we don't need it.

knatten replied:
> I would not expect that calling transform would modify the frame.

I too prefer functions which don't modify the object, but rather return a new, modified object. But many users probably won't need the original non-transformed point cloud again, and I think (didn't check) in-place transforms can be done slightly faster. In my opinion, it would be nice if we had both an in-place `transform()` and a `transformed() const` which returns a new point cloud.

SatjaSivcev reviewed (changes requested):
Changes required

knatten commented:
Did you accidentally add a lot of build artifacts here?

SatjaSivcev commented on source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.cs:
```diff
@@ -0,0 +1,157 @@
+                    var transformMatrix = mathDotNetToZivid(transformBaseToCamera);
+                    pointCloud.Transform(transformMatrix);
+
+                    var saveFile = "ZividGemTransformed.zdf";
```
The var name is not consistent with what we use in other samples. I think we use fileName don't we?

chrisasc replied:
We use dataFile, but I already use that var name when reading point cloud, hence I need a new name for the saved file. I can change it til fileName instead of saveFile

SatjaSivcev replied:
I don't like 'saveFile' - it sounds like imperative. How about dataFile and dataFileTransformed / dataTransformedFile / transformedDataFile?

chrisasc replied:
That is a better solution! I like the name "dataFileTransformed" as it resonates with what we call the file, ZividGem -> ZividGemTransformed and dataFile -> dataFileTransformed

SatjaSivcev replied:
Agree, still to be implemented.

chrisasc replied:
done

chrisasc commented:
> Did you accidentally add a lot of build artifacts here?

I was not sure about what to do with the build artifacts (the entire packages folder actually), so I thought it was better to add it and then remove it than the other way around. This is the first sample in C# where we depend on some third-party libraries.

knatten commented:
> I was not sure about what to do with the build artifacts (the entire packages folder actually), so I thought it was better to add it and then remove it than the other way around. This is the first sample in C# where we depend on some third-party libraries.

Adding these to the repo sounds like the wrong approach. Did you consider using a pacakge manager such as nuget?

chrisasc commented:
> > I was not sure about what to do with the build artifacts (the entire packages folder actually), so I thought it was better to add it and then remove it than the other way around. This is the first sample in C# where we depend on some third-party libraries.
> 
> Adding these to the repo sounds like the wrong approach. Did you consider using a package manager such as nuget?

Yes, I will write how to install the two packages (using nugget) in the readme. This way I dont need any of the artifact dependencies

chrisasc commented:
Ready for next round!

SatjaSivcev commented on README.md:
````diff
@@ -45,9 +45,41 @@ Note: The version tested with Zivid cameras is 2.0.0.
+```
+Open *ZividNETSamples.sln* in Visual Studio, build it, and run one if the samples.
+
+## External libaries 
+Some of the samples depend on external libraries, in particular *YamlDotNet* and *MathNet.Numerics*. Both of these libraries can be installed using NuGet package manager within Visual Studio. 
+
+To install the libraries:
+- **YamlDotNet**: Type `PM> Install-Package YamlDotNet` in the Package Manager Console.           
+- **MathNet.Numerics**: Type `PM> Install-Package MathNet.Numerics` in the Package Manager Console.
````
This is nice but in converges from what we have for C++ repo. In C++ repo, we just have links to the dependencies, the instructions on how to install them are on the KB. For C#, you've added the instructions here. I think we should be consistent.

chrisasc replied:
(reffering to Anders comment), Using NuGet packet manager this is done automatically, hence we don't need any explanation on how to download these libraries. After I have made NuGet work/understand how it works, I can create some instructions on KB (I saw that we don't have instructions on how to install and build Zivid samples in Ubuntu either, so can create that as well).

SatjaSivcev replied:
I don't think C#/.NET works on Ubuntu

SatjaSivcev replied:
But do add instructions on KB for packages if needed

knatten replied:
There should be no documentation needed. If installing these requirements require any manual effort from the user, we're doing it wrong.

chrisasc replied:
Done, removed all Ubuntu related stuff and updated the KB article on how to install third-party libraries; https://zivid.atlassian.net/wiki/spaces/ZividKB/pages/534708242/Build+C+Samples+using+Visual+Studio

knatten replied:
LGTM, but leaving it to @SatjaSivcev to resolve since it's his thread.

SatjaSivcev replied:
https://support.zivid.com/latest/academy/samples/c-sharp/build-c-sharp-samples-using-visual-studio.html

SatjaSivcev reviewed (changes requested):
More changes needed. Don't forget to check the comments from the initial review, some of the issues are not resolved yet.

knatten reviewed (changes requested):
Haven't look at the actual code yet, just some overall comments.

You should also clean up the commit history so that every commit is "true" on its own . Instead of having one commit saying "Update readme", and then another commit that does the thing you updated the readme for, you should have one commit which both does the thing and updates the readme. This ensures that no matter which version of the repo someone checks out, everything builds and is documented correctly. Also, if we later revert a commit, we don't suddenly have wrong docs, stuff that doesn't build etc.

I also suggest you make one separate commit for moving the old sample and one for adding the new one. It makes it easier to see what's going on when you aren't doing two things at the same time.

knatten commented on README.md:
```diff
@@ -30,7 +30,12 @@ There are two main categories of samples: **Camera** and **Applications**. The s
+    - [**HandEyeCalibration**][HandEyeCalibrationReadme-url]
+        - [**HandEyeCalibration**][HandEyeCalibration-url] - Perform Hand-Eye calibration.
+        - [**UtilizeEyeInHandCalibration**][UtilizeEyeInHandCalibration-url] - Transform single data point or entire point cloud from camera frame to robot base frame using Eye-in-Hand calibration matrix.
+          - **Dependencies:**
```
Did you consider using NuGet for this? Then:
- Everything would just work automatically for the user, they won't have to install anything manually
- They'll automatically get the correct version
- You don't need to document which dependencies are needed and where to get them (and this documentation can't accidentally get outdated)
- Packages get installed locally to the project, users don't need to install things globally on their system
- It will be possible to build in CI since we don't need to manually install stuff first
- This is how .NET users will expect things to work

chrisasc replied:
> Did you consider using NuGet for this?

That would be the ideal solution! How do I use NuGet this way? 
Is it enough to have package.config file in the sample folder? 

(Sorry for asking stupid questions, but i'm not familiar with C# and this is the first sample where we rely on some third-party packages)

knatten replied:
> If i have the package.config in the sample folder, they will automatically be downloaded by NuGet when I build the sample. This was very easy and required no manual work from me (as a user). Is this what you mean?

That is what I mean. If you add these dependencies using NuGet in Visual Studio, it will be set up to work like this automatically. When a user then clones this repo and builds using Visual Studio, they don't have to do anything manually.
 
> This is only possible in Windows I think, but I will need to check it out.

Our .NET api is Windows only, you should remove everything you write about Ubuntu from this PR.

chrisasc replied:
done

chrisasc commented on README.md:
```diff
@@ -40,9 +45,41 @@ Note: The version tested with Zivid cameras is 2.0.0.
+
+Open *ZividNETSamples.sln* in Visual Studio, build it, and run it. If you are uncertain about doing this, check out our [**tutorials for configuring and building C# Samples**](https://zivid.atlassian.net/wiki/spaces/ZividKB/pages/534708242).
+
+### Ubuntu
```
I don't know yet, need to test that out before merging to master. I suspect that NuGet packet manager automatically downloads packages is for Windows only.

SatjaSivcev replied:
I don't think we support C# on Ubuntu?

SatjaSivcev replied:
Is that even possible?

knatten replied:
> I don't think we support C# on Ubuntu?

The Zivid .NET api is Windows only.

SatjaSivcev replied:
Exactly, so this should be removed.

chrisasc replied:
Done, removed

chrisasc commented on source/Applications/Advanced/HandEyeCalibration/README.md:
```diff
@@ -0,0 +1,34 @@
+# Hand Eye Calibration
+
+To fully understand Hand-Eye Calibration, please see the [tutorial][Tutorial-url] in our Knowledge Base.
+
```
Added a sentence explaining that the following applications creates a Transformation Matrix from data provided by a user

SatjaSivcev replied:
Please create an issue to add this to the C++ repo as well

SatjaSivcev replied:
Please check if this has been done, if not, do it.

knatten replied:
I'm ok with resolving this now, don't know if @SatjaSivcev wants to keep it open until the C++ thing is fixed?

chrisasc replied:
This is resolved both in C++ and in C# with the following text in the readme: 
![image](https://user-images.githubusercontent.com/25568033/131314562-d8c49f54-2d6b-4f57-9405-bb9060254b69.png)

chrisasc commented on source/Applications/Advanced/HandEyeCalibration/README.md:
```diff
@@ -0,0 +1,34 @@
+   4. Repeat i.-iii. until 10-20 pose pairs are collected
+   5. Enter command to perform calibration and return a **Transformation Matrix**
+
+[ZividHandEyeCalibration][ZividHandEyeCalibration-url]
```
I agree. We have the same in C++, but it is a bit strange to have this link here (since its not a sample). @SatjaSivcev, do you know why it was added here in the first place? I think we either should remove it or expose the code as a proper sample.

SatjaSivcev replied:
@torbsorb was the one who added it. It was written as a tutorial so I think it should stay there. Perhaps we should clarify that it comes with the installer for Win and is part of tools deb?

knatten replied:
> @torbsorb was the one who added it. It was written as a tutorial so I think it should stay there. Perhaps we should clarify that it comes with the installer for Win and is part of tools deb?

Good idea

chrisasc replied:
Included a sentence there I say that this sample comes with the installer in Windows and is part of tools deb in Ubuntu

chrisasc commented:
Fixed commit and commit history and rebased on master. Ready for next round

SatjaSivcev commented:
Looks good to me @chrisasc

I would squash into one commit called 'Add __ sample and move __ sample'.

Have you double checked that it works well?

chrisasc commented:
> Looks good to me @chrisasc
> 
> I would squash into one commit called 'Add __ sample and move __ sample'.
> 
> Have you double checked that it works well?

I separated them into two commits because of Anders comments; _"I also suggest you make one separate commit for moving the old sample and one for adding the new one. It makes it easier to see what's going on when you aren't doing two things at the same time."_
but i can squash them together. 

I've tested them and they work on my computer

SatjaSivcev commented:
> 
> 
> > Looks good to me @chrisasc
> > I would squash into one commit called 'Add __ sample and move __ sample'.
> > Have you double checked that it works well?
> 
> I separated them into two commits because of Anders comments; _"I also suggest you make one separate commit for moving the old sample and one for adding the new one. It makes it easier to see what's going on when you aren't doing two things at the same time."_
> but i can squash them together.
> 
> I've tested them and they work on my computer

I see. Having two commits works just as fine. However, in that case I would not have this third commit 'Update main readme and add hand-eye folder readme' but rather have those (appropriate) code changes in the first two commits.
The point is that each commit does one complete change. As you have it now, the third commit is just something that should have been done in the first too.

chrisasc commented:
> I see. Having two commits works just as fine. However, in that case I would not have this third commit 'Update main readme and add hand-eye folder readme' but rather have those (appropriate) code changes in the first two commits.
> The point is that each commit does one complete change. As you have it now, the third commit is just something that should have been done in the first too.

That makes sense, I squashed the readme and add UtilizeH... into one commit.

SatjaSivcev commented on README.md:
```diff
@@ -30,7 +30,12 @@ There are two main categories of samples: **Camera** and **Applications**. The s
       - [**ZDF2PLY**][ZDF2PLY-url] - Convert point cloud from ZDF a file to a PLY file.
   - **Advanced**
-    - [**HandEyeCalibration**][HandEyeCalibration-url] - Perform Hand-Eye calibration.
+    - [**HandEyeCalibration**][HandEyeCalibrationReadme-url]
+        - [**HandEyeCalibration**][HandEyeCalibration-url] - Perform Hand-Eye calibration.
```
Can you separate out these two lines of code from this and add them to the previous commit. This is basically showing moving the sample.

chrisasc replied:
done

SatjaSivcev commented on README.md:
```diff
@@ -67,6 +82,10 @@ Zivid Samples are distributed under the [BSD license](LICENSE).
 [ZDF2PLY-url]: source/Applications/Basic/FileFormats/ZDF2PLY/ZDF2PLY.cs
-[HandEyeCalibration-url]: source/Applications/Advanced/HandEyeCalibration/HandEyeCalibration.cs
+[HandEyeCalibrationReadme-url]: source/Applications/Advanced/HandEyeCalibration/README.md
+[HandEyeCalibration-url]: source/Applications/Advanced/HandEyeCalibration/HandEyeCalibration/HandEyeCalibration.cs
```
Can you separate out this line of code from this and add it to the previous commit. This is basically showing moving the sample.

chrisasc replied:
done

SatjaSivcev commented on README.md:
```diff
@@ -67,6 +82,10 @@ Zivid Samples are distributed under the [BSD license](LICENSE).
+[UtilizeEyeInHandCalibration-url]: source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.cs
 [MultiCameraCalibration-url]: source/Applications/Advanced/MultiCameraCalibration/MultiCameraCalibration.cs
 [Downsample-url]: source/Applications/Advanced/Downsample/Downsample.cs
+
+
```
delete these lines?

chrisasc replied:
done

SatjaSivcev reviewed (changes requested):
Minor changes required.

SatjaSivcev reviewed (approved):
This looks good to me. I think that @knatten should approve as well before merging.

knatten commented:
There's one commit that says "Move HandEyeCalibration sample to new folder" but also does some seemingly unrelated updates to the README. I suggest splitting this one in two.

knatten commented on README.md:
```diff
@@ -41,9 +46,19 @@ Note: The version tested with Zivid cameras is 2.1.0.
+
+Open *ZividNETSamples.sln* in Visual Studio, build it, and run it. If you are uncertain about doing this, check out our [**tutorials for configuring and building C# Samples**](https://zivid.atlassian.net/wiki/spaces/ZividKB/pages/534708242).
+
+Some of the samples depend on external libraries, in particular *YamlDotNet* and *MathNet.Numerics*. These libaries will be installed automatrically through NuGet when building the sample.
```
You add this in a commit which happens before these dependencies are actually used. So if we for instance were to keep the commit that updates the readme, but revert the one that adds the new sample, the readme will be wrong. I recommend always updating the readme in the same commit that introduces a particular change.

chrisasc replied:
Done, changed such that README is changed in the `AddUtilizeEyeInHandSample` commit.

knatten commented on README.md:
```diff
@@ -41,9 +46,19 @@ Note: The version tested with Zivid cameras is 2.1.0.
+
+Open *ZividNETSamples.sln* in Visual Studio, build it, and run it. If you are uncertain about doing this, check out our [**tutorials for configuring and building C# Samples**](https://zivid.atlassian.net/wiki/spaces/ZividKB/pages/534708242).
+
+Some of the samples depend on external libraries, in particular *YamlDotNet* and *MathNet.Numerics*. These libaries will be installed automatrically through NuGet when building the sample.
```
You should spell check this file. I recommend enabling the spell checker in whichever editor you use.

chrisasc replied:
done

knatten commented on README.md:
```diff
@@ -30,7 +30,12 @@ There are two main categories of samples: **Camera** and **Applications**. The s
+    - [**HandEyeCalibration**][HandEyeCalibrationReadme-url]
+        - [**HandEyeCalibration**][HandEyeCalibration-url] - Perform Hand-Eye calibration.
+        - [**UtilizeEyeInHandCalibration**][UtilizeEyeInHandCalibration-url] - Transform single data point or entire point cloud from camera frame to robot base frame using Eye-in-Hand calibration matrix.
+          - **Dependencies:**
```
I suggest mentioning here that the user does not need to install this manually, or they'll maybe start doing that before the discover that it just works out of the box.

chrisasc replied:
done

knatten commented on source/ZividNETSamples.sln:
```diff
@@ -31,12 +31,14 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CaptureFromFileVis3D", "App
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CaptureVis3D", "Applications\Basic\Visualization\CaptureVis3D\CaptureVis3D.csproj", "{68EB8781-17AA-42E9-9524-0C18A45278C3}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "HandEyeCalibration", "Applications\Advanced\HandEyeCalibration\HandEyeCalibration.csproj", "{8ACFFEC6-34E5-4E14-B08C-60051B1CFD2F}"
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "HandEyeCalibration", "Applications\Advanced\HandEyeCalibration\HandEyeCalibration\HandEyeCalibration.csproj", "{8ACFFEC6-34E5-4E14-B08C-60051B1CFD2F}"
```
It looks like this was changed in the commit which adds `UtilizeEyeInHandCalibration`, rather than in the one that moves `HandEyeCalibration`. So the commit that moves `HandEyeCalibration` probably doesn't build now, you should check that.

chrisasc replied:
Done. Moved it to the commit that moves HE sample

knatten reviewed (changes requested):
Did an initial review of the surrounding stuff now, will start looking at the actual sample next. Just submitting this in the mean time so you can start looking at it.

knatten commented:
I think there's something wrong in the sample data. I've installed the sample data according to https://zivid.atlassian.net/wiki/spaces/ZividKB/pages/450363393/Sample+Data. When I try to run the sample, I get this error message:

```
Reading camera pose in end-effector frame (result of eye-in-hand calibration)
Reading end-effector pose in robot base frame
Error: (Line: 1, Col: 1, Idx: 0) - (Line: 1, Col: 6, Idx: 5): While scanning a directive, found unexpected non-alphabetical character.
Press any key to continue . . .
```

The exception happens the first time I reach `yaml.Load(reader);` in `readTransform`.  The `transformFile` argument is `C:\ProgramData/Zivid/EyeInHandTransform.yaml`, and the content of that file is

```
%YAML:1.0
---
PoseState: !!opencv-matrix
   rows: 4
   cols: 4
   dt: f
   data: [ -1.52643966e-02, -9.99315023e-01, 3.37116420e-02,
       -2.68839798e+01, -9.65478182e-01, 5.96248358e-03, -2.60415733e-01,
       6.69086380e+01, 2.60036349e-01, -3.65229435e-02, -9.64907885e-01,
       -3.33924942e+01, 0., 0., 0., 1. ]
```

I don't know much about yaml, but when I try to paste that into http://www.yamllint.com/, it says the file is not valid yaml.

Does this work on your machine at all?

chrisasc commented:
> I think there's something wrong in the sample data. I've installed the sample data according to https://zivid.atlassian.net/wiki/spaces/ZividKB/pages/450363393/Sample+Data. When I try to run the sample, I get this error message:
> 
> ```
> Reading camera pose in end-effector frame (result of eye-in-hand calibration)
> Reading end-effector pose in robot base frame
> Error: (Line: 1, Col: 1, Idx: 0) - (Line: 1, Col: 6, Idx: 5): While scanning a directive, found unexpected non-alphabetical character.
> Press any key to continue . . .
> ```
> 
> The exception happens the first time I reach `yaml.Load(reader);` in `readTransform`. The `transformFile` argument is `C:\ProgramData/Zivid/EyeInHandTransform.yaml`, and the content of that file is
> 
> ```
> %YAML:1.0
> ---
> PoseState: !!opencv-matrix
>    rows: 4
>    cols: 4
>    dt: f
>    data: [ -1.52643966e-02, -9.99315023e-01, 3.37116420e-02,
>        -2.68839798e+01, -9.65478182e-01, 5.96248358e-03, -2.60415733e-01,
>        6.69086380e+01, 2.60036349e-01, -3.65229435e-02, -9.64907885e-01,
>        -3.33924942e+01, 0., 0., 0., 1. ]
> ```
> 
> I don't know much about yaml, but when I try to paste that into http://www.yamllint.com/, it says the file is not valid yaml.
> 
> Does this work on your machine at all?

It says that it is not valid because of the ":".
I did some digging on this. I was not able to read the yaml file either (saw that I had modified it a bit locally). The reason is that yamlDotNet supports Yaml v1.1 and v1.2, but not v.1.0. In v.1.0 the correct syntax is "%YAML: 1.0", while in the newer versions its "%YAML 1.1" (without the :). 

I also found out that OpenCV uses v1.1 when creating .yaml files. That is why all of our robot-related .yaml files start with 
> %YAML:1.0
> ---

OpenCV saves files as 1.0, but it can read yaml files from of all versions.
YamlDotNet can only read 1.1 and 1.2 yaml files.

chrisasc commented:
Ready for next round.

knatten commented:
The first commit says "Restructure: Move HandEyeCalibration sample to new folder". It seems it doesn't move the sample though, but duplicates it. It also introduces a README file which is not mentioned in the commit message, and which describes a new `UtilizeEyeInHandCalibration` sample which does not exist yet in that commit.

Can you please fix this and force push? Then ping me so I can get started on the proper review (a force push in the middle of a review is often cumbersome, so I'd rather wait with the details to after the force push).

chrisasc commented:
Back to this PR again!

Changes since the last review:
- Updated the code so that it accepts both versions of YAML files. (Same as how Robert does it in https://github.com/zivid/zivid-csharp-samples/pull/34).
- Updated links to our new KB.
- Fixed commit messages so that the first one only moves the existing HE sample and the second commit adds this new same.  


Ready for another round of review. The sample has been tested on Windows 10 with VS 2019.

knatten commented on README.md:
````diff
@@ -53,9 +58,19 @@ Note: The samples require Zivid SDK v2 (minor version 2.2 or newer).
+git clone https://github.com/zivid/zivid-csharp-samples
+```
+
+Open *ZividNETSamples.sln* in Visual Studio, build it and run it. If you are uncertain about doing this, check out our [**tutorials for configuring and building C# Samples**](https://zivid.atlassian.net/wiki/spaces/ZividKB/pages/534708242).
````
Is this link still valid, or should it point to the new kb now?

chrisasc replied:
No, corrected with the new KB

knatten commented on source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.cs:
```diff
@@ -0,0 +1,159 @@
+            var endEffectorPose = getTransformationMatrixFromYAML(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData)
+                               + "/Zivid/RobotTransform.yaml");
+
+            // Converting to Mat.NET matrices for easier computation
```
Nitpick: I think you mean Math.NET

chrisasc replied:
You are correct!

knatten commented on source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.cs:
```diff
@@ -0,0 +1,159 @@
+            return zividMatrix;
+        }
+    }
+    static Matrix<float> zividToMathDotNet(float[,] zividMatrix)
```
I really like how you introduce named functions for these very simple one-liners. It makes the code which uses them easier to understand.

chrisasc replied:
Thanks!

knatten commented:
I'm not much involved in samples these days, so I'll leave it up to someone from @zivid/customer-success to approve of what the sample does and how it does it, how it fits in the big picture, what the documentation is like etc. I added a few tiny comments, but otherwise this seems good from my point of view now. There's an unrelated update to README.md which could have been a separate commit, but I guess it's unlikely that we'll revert this anyway so I don't really mind it.

I have not tested this again since last time btw, so someone in @zivid/customer-success other than @chrisasc should probably do that.

SatjaSivcev commented on source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.csproj:
```diff
@@ -0,0 +1,94 @@
+    <DebugSymbols>true</DebugSymbols>
+    <DebugType>full</DebugType>
+    <Optimize>false</Optimize>
+    <OutputPath>..\..\..\..\build\$(Configuration)\$(Platform)</OutputPath>
```
You probably need one more ..\

chrisasc replied:
done

SatjaSivcev commented on source/Applications/Advanced/HandEyeCalibration/UtilizeEyeInHandCalibration/UtilizeEyeInHandCalibration.csproj:
```diff
@@ -0,0 +1,94 @@
+    <PlatformTarget>x64</PlatformTarget>
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>..\..\..\..\build\$(Configuration)\$(Platform)</OutputPath>
```
You probably need one more ..\

chrisasc replied:
done

SatjaSivcev commented:
> I'm not much involved in samples these days, so I'll leave it up to someone from @zivid/customer-success to approve of what the sample does and how it does it, how it fits in the big picture, what the documentation is like etc. I added a few tiny comments, but otherwise this seems good from my point of view now. There's an unrelated update to README.md which could have been a separate commit, but I guess it's unlikely that we'll revert this anyway so I don't really mind it.
> 
> I have not tested this again since last time btw, so someone in @zivid/customer-success other than @chrisasc should probably do that.

I have reviewed it and it looks good to me. I added a small fixup commit to fix the build path. Rebase, squash, merge @chrisasc
